> Build a __regression model__ using the deep learning Keras library,  
    and then you will experiment with increasing the number of training epochs and changing number of hidden layers   
    and you will see how changing these parameters impacts the performance of the model.  

The breakdown will be:

Part A: 5 marks

Part B: 5 marks

Part C: 5 marks

Part D: 5 marks  

>Model the same data about concrete compressive strength:  
[data_](https://cocl.us/concrete_data)

## Part A: Build a baseline Model   

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the 
train_test_split
helper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

In [1]:
# import pandas and numpy libraries
import pandas as pd
import numpy as np 

### import data

In [2]:
#import from provided URL

url = r'https://cocl.us/concrete_data'
df =pd.read_csv(url)
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer   
0   540.0                 0.0      0.0  162.0               2.5  \
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

### 1. Split data

In [3]:
# split to predictors (X) and Target (y) variables

X = df[df.columns[df.columns!="Strength"]]
y = df['Strength'] 

In [4]:
# split the data to training and test sets randomly with a test size of 30%
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state=101)

print("Size of Sets\nX_train size:", X_train.shape[0],"\ny_train size:",y_train.shape[0],
      "\n\nX_test size:", X_test.shape[0],"\ny_test size:", y_test.shape[0])

Size of Sets
X_train size: 721 
y_train size: 721 

X_test size: 309 
y_test size: 309


#### Get the number of predictors for the model

In [5]:
# number of predictors
no_pred = X.shape[1]
no_pred

8

##### Build the baseline model

In [6]:
#import keras library
import keras
#import packages
from keras.models import Sequential
from keras.layers import Dense,Input

In [7]:
def reg_model():
    #create regression model
    model = Sequential()
    model.add(Input(shape=(no_pred,)))
    model.add(Dense(50,activation='relu'))
    model.add(Dense(50,activation='relu'))
    model.add(Dense(1))

    # compile the model
    model.compile(optimizer='adam', loss = 'mean_squared_error')
    return model


### 2. Train with 50 epochs

In [8]:
# fit the regression model 
model=reg_model()
model.fit(x= X_train ,y =y_train, epochs=50,verbose=0)

### 3. Evaluate the model on the test data and compute the mean squared error

In [9]:
from sklearn.metrics import mean_squared_error
# evalutae using MSE
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mse

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


54.0465962881772

### 4. create a list of 50 mean squared errors.

In [10]:
# Create a list of 50 MSEs 
mse_list =[]
# using a for loop for this task
for i in range(50):
    model=reg_model()
    model.fit(X_train,y_train,epochs=50,verbose=0)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list.append(mse)
    print(f"\n\nIteration {i+1}, Mean Squared Error: {mse}")

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 1, Mean Squared Error: 50.75515831253314
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 2, Mean Squared Error: 49.842255823386395
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 3, Mean Squared Error: 50.17140312313064
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step


Iteration 4, Mean Squared Error: 59.46539330430616
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 5, Mean Squared Error: 73.40639174980222
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 6, Mean Squared Error: 52.86874861307732
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 7, Mean Squared Error: 45.80261988242409
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step


Iteration 8, Mean Squared Error: 75.49250935937117
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 9, Mean Squared Error: 63.38769309034982
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step


Iteration 10, Mean Squared Error: 59.99500360761787
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step


Iteration 11, Mean Squared

### 5. Report the mean and the standard deviation of the mean squared errors.

In [11]:
#from the list of mse, get the mean std and mean of squared errors
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

Mean of Mean Squared Errors: 63.47559001131865
Standard Deviation of Mean Squared Errors: 13.56583306718807


## Part B. Normalize the data  
Repeat Part A but use a normalized version of the data.   
Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

In [12]:
# split to predictors (X) and Target (y) variables

X = df[df.columns[df.columns!="Strength"]]
y = df['Strength'] 

In [13]:
# Normalise the data
X_norm = (X - X.mean()) / X.std()
X_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer   
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147  \
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [14]:
# split the data to training and test sets randomly with a test size of 30%
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_norm,y,test_size=0.3,random_state=101)

print("Size of Sets\nX_train size:", X_train.shape[0],"\ny_train size:",y_train.shape[0],
      "\n\nX_test size:", X_test.shape[0],"\ny_test size:", y_test.shape[0])

Size of Sets
X_train size: 721 
y_train size: 721 

X_test size: 309 
y_test size: 309


In [15]:
def reg_model():
    #create regression model
    model = Sequential()
    model.add(Input(shape=(no_pred,)))
    model.add(Dense(50,activation='relu'))
    model.add(Dense(50,activation='relu'))
    model.add(Dense(1))

    # compile the model
    model.compile(optimizer='adam', loss = 'mean_squared_error')
    return model

In [16]:
# fit the regression model 
model=reg_model()
model.fit(x= X_train ,y =y_train, epochs=50,verbose=0)

In [17]:
from sklearn.metrics import mean_squared_error
# evalutae using MSE
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mse

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step


123.34147215512438

In [18]:
# Create a list of 50 MSEs 
mse_list =[]
# using a for loop for this task
for i in range(50):
    model=reg_model()
    model.fit(X_train,y_train,epochs=50,verbose=0)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list.append(mse)
    print(f"\n\nIteration {i+1}, Mean Squared Error: {mse}")

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 1, Mean Squared Error: 118.82151516862136
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 2, Mean Squared Error: 129.6048205847144
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 3, Mean Squared Error: 118.71603061843751
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 4, Mean Squared Error: 104.59646139046383
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step


Iteration 5, Mean Squared Error: 132.5114313290645
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 6, Mean Squared Error: 115.10014396557033
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 7, Mean Squared Error: 110.59549409326995
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step


Iteration 8, Mean Squared Error: 104.43382148258206
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 9, Mean Squared Error: 113.56521527162306
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 10, Mean Squared Error: 109.96441219531712
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step


Iteration 11, Mean 

In [19]:
#from the list of mse, get the mean std and mean of squared errors
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

Mean of Mean Squared Errors: 117.94445744576639
Standard Deviation of Mean Squared Errors: 8.304650964976304


**How does the mean of the mean squared errors compare to that from Step A?**

 * The mean of the squared errors (MSE) for Step B is higher than for Step A, indicating worse performance in terms of average squared error.

 * The errors in Step A are more dispersed (higher standard deviation) than in Step B, where errors are more consistent (lower standard deviation).
>In conclusion, Step A shows a better performance with a lower MSE but has more variability in the errors, while Step B has a higher MSE with less variability in the errors.

## Part C. Increase the number of epochs   
Repeat Part B but use 100 epochs this time for training.

In [20]:
# split to predictors (X) and Target (y) variables

X = df[df.columns[df.columns!="Strength"]]
y = df['Strength'] 

In [21]:
# Normalise the data
X_norm = (X - X.mean()) / X.std()
X_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer   
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147  \
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [22]:
# split the data to training and test sets randomly with a test size of 30%
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_norm,y,test_size=0.3,random_state=101)

print("Size of Sets\nX_train size:", X_train.shape[0],"\ny_train size:",y_train.shape[0],
      "\n\nX_test size:", X_test.shape[0],"\ny_test size:", y_test.shape[0])

Size of Sets
X_train size: 721 
y_train size: 721 

X_test size: 309 
y_test size: 309


In [23]:
def reg_model():
    #create regression model
    model = Sequential()
    model.add(Input(shape=(no_pred,)))
    model.add(Dense(50,activation='relu'))
    model.add(Dense(50,activation='relu'))
    model.add(Dense(1))

    # compile the model
    model.compile(optimizer='adam', loss = 'mean_squared_error')
    return model

In [24]:
# Create a list of 50 MSEs 
mse_list =[]
# using a for loop for this task
for i in range(50):
    model=reg_model()
    model.fit(X_train,y_train,epochs=100,verbose=0)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list.append(mse)
    print(f"\n\nIteration {i+1}, Mean Squared Error: {mse}")

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 1, Mean Squared Error: 46.61291236097073
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step


Iteration 2, Mean Squared Error: 47.076283659060756
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step


Iteration 3, Mean Squared Error: 46.757999946957476
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step


Iteration 4, Mean Squared Error: 46.031332427152634
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 5, Mean Squared Error: 48.122504599005765
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step


Iteration 6, Mean Squared Error: 46.01635117466323
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 7, Mean Squared Error: 50.84267878949098
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step


Iteration 8, Mean Squared Error: 44.82560063676471
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step


Iteration 9, Mean Squared Error: 46.69922443492271
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step


Iteration 10, Mean Squared Error: 46.16818817057824
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step


Iteration 11, Mean Squa

In [25]:
#from the list of mse, get the mean std and mean of squared errors
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

Mean of Mean Squared Errors: 47.15033297547117
Standard Deviation of Mean Squared Errors: 2.5021523771063805


**How does the mean of the mean squared errors compare to that from Step B**

* The mean MSE in Step C (47.15) is significantly lower than in Step B (117.94). This indicates that, on average, the squared errors are much smaller in Step C, suggesting better model performance.

* The standard deviation of MSE in Step C (2.50) is also lower than in Step B (8.30). This suggests that the MSEs are more consistent (less variable) in Step C compared to Step B.

> Step C outperforms Step B significantly in terms of both the average performance (lower mean MSE) and the consistency of the performance (lower standard deviation of MSE).

## Part D. Increase the number of hidden layers

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

In [26]:
# split to predictors (X) and Target (y) variables

X = df[df.columns[df.columns!="Strength"]]
y = df['Strength'] 

In [27]:
# split the data to training and test sets randomly with a test size of 30%
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_norm,y,test_size=0.3,random_state=101)

print("Size of Sets\nX_train size:", X_train.shape[0],"\ny_train size:",y_train.shape[0],
      "\n\nX_test size:", X_test.shape[0],"\ny_test size:", y_test.shape[0])

Size of Sets
X_train size: 721 
y_train size: 721 

X_test size: 309 
y_test size: 309


In [28]:
# Normalise the data
X_norm = (X - X.mean()) / X.std()
X_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer   
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147  \
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [29]:
def reg_model():
    #create regression model
    model = Sequential()
    model.add(Input(shape=(no_pred,)))
    model.add(Dense(10,activation='relu'))
    model.add(Dense(10,activation='relu'))
    model.add(Dense(10,activation='relu'))
    model.add(Dense(1))

    # compile the model
    model.compile(optimizer='adam', loss = 'mean_squared_error')
    return model

In [30]:
# Create a list of 50 MSEs 
mse_list =[]
# using a for loop for this task
for i in range(50):
    model=reg_model()
    model.fit(X_train,y_train,epochs=50,verbose=0)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list.append(mse)
    print(f"\n\nIteration {i+1}, Mean Squared Error: {mse}")


10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 1, Mean Squared Error: 141.28841290803803
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step


Iteration 2, Mean Squared Error: 131.58813263838448
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step


Iteration 3, Mean Squared Error: 151.82446079369896
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step


Iteration 4, Mean Squared Error: 153.70577264818087
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step


Iteration 5, Mean Squared Error: 158.6691970902952
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step


Iteration 6, Mean Squared Error: 151.75469542770443
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step


Iteration 7, Mean Squared Error: 147.74788300439366
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step


Iteration 8, Mean Squared Error: 158.18475227542586
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


Iteration 9, Mean Squared Error: 141.7304079340366
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step


Iteration 10, Mean Squared Error: 143.21346238425312
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step


Iteration 11, Mean 

In [31]:
#from the list of mse, get the mean std and mean of squared errors
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

Mean of Mean Squared Errors: 141.59435346607648
Standard Deviation of Mean Squared Errors: 14.669533346464172


**How does the mean of the mean squared errors compare to that from Step B?**

* Part D shows worse performance in terms of the mean MSE compared to Part B, with a higher mean MSE indicating larger average squared errors.
* The higher standard deviation of MSE in Part D also indicates that the errors are more variable compared to Part B, suggesting less consistency in the squared errors.
> Part B performs better than Part D in terms of both the average performance (lower mean MSE) and the consistency of the performance (lower standard deviation of MSE).
